source/common/upstream/health_checker_impl.cc

```diff
@@ -723,6 +724,7 @@ void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::onInterval() {
       parent_.service_method_.name());
   headers_message->headers().insertUserAgent().value().setReference(
       Http::Headers::get().UserAgentValues.EnvoyHealthChecker);
+  Router::FilterUtility::setUpstreamScheme(headers_message->headers(), *parent_.cluster_.info());
```

Comment:
Since this is required in gRPC spec, maybe add this to `Grpc::Common::prepareHeaders`?

Reply:
Yeah, it's a reasonable option. And since `scheme` header is also required in h2 spec, we might have another place to add this to ensure all out-going h2 requests have necessary headers. Still I couldn't find the proper place.

Reply:
With some digging the code base, other gRPC client codes in Envoy use a HttpAsyncClient which use a Router. Since a router sets proper h2 headers, it seems other gRPC client codes don't need to inject the `scheme` header. This has been commented https://github.com/envoyproxy/envoy/issues/2583#issuecomment-364789736.